### ISSUE TITLE
Refresh issues lead to show artifacts/hidden elements on Wayland

### ISSUE THREAD

**Issue by Digitalone1** [wxWidgets/wxWidgets#25036]
### Description

Hello everyone. First of all, sorry if a similar issue has been already reported, but I didn't find anything.

I open this ticket following the problem reported here by another user: https://forum.filezilla-project.org/viewtopic.php?t=61130

The app is Filezilla which uses wxwidgets. I run it on Linux, Gnome Wayland desktop. Maybe it's a Filezilla issue, but the problem is resolved using the X11 backend, so I though to report it here.

The issue is that using Filezilla in default mode (which I assume uses native Wayland) I have artifacts on the screen when the app is supposed to refresh the items/widgets. See the [video](https://youtu.be/nE3nPmFWQ5A) made by the user who reported the issue on the Filezilla forum.

The workaround is to launch the app with `GDK_BACKEND=x11` env variable, which I assume does not use native Wayland, but X11 backend.  

#### Bug description:

I made a video of the issue on my system.

https://github.com/user-attachments/assets/8389ee04-77e7-4968-9dec-7e34bbe403bb

First of all, there are artifacts at the bottom of the list which I'm not able to clear.

At the beginning, as you can see, the list is empty, but in reality the directory contains 2 subfolders which are not shown when I entered. You can only see them when I select them with the mouse.

Then I upload one folder and you see that the bottom message is still shown even if the upload is completed. It's supposed to be hidden, but it's still there. The only way to remove it, is to update the folder.

At last, I go back to the parent folder, but the list is shown another time as empty while, in reality, there are lots of folders.

#### Expected vs observed behaviour:

The expected behavior is that the app refreshes itself correctly in order to show the proper items.

### Platform and version information

Gnome Desktop v47 on Arch Linux. [Filezilla](https://archlinux.org/packages/extra/x86_64/filezilla/) is built on [wxwidgets-gtk3](https://archlinux.org/packages/extra/x86_64/wxwidgets-gtk3/) and [wxwidgets-common](https://archlinux.org/packages/extra/x86_64/wxwidgets-common/) v3.2.6.

**Comment by vadz** [comment]
If the control with the items is `wxListCtrl` (and not `wxDataViewCtrl`), then this looks very similar to #24443 and if FileZilla uses `wxEVT_IDLE` for refreshing, this must definitely be the same thing (and it still might be even if it doesn't).

Unfortunately I have no idea about the artifacts at the bottom.

**Comment by Digitalone1** [comment]
> Unfortunately I have no idea about the artifacts at the bottom.

They are not refreshed from the a previous list.

I have this issue from more than a year. Do you think it's related to Gnome/Gtk or is it likely happening also on other Wayland compositors?

**Comment by vadz** [comment]
I think it's our bug (at least if it's really the same thing as the other one), but, somehow, I can't reproduce it reliably in a simple example. I'll try to debug it again...

FWIW I see it with Sway, so I'm pretty sure it's not compositor-related.

**Comment by Digitalone1** [comment]
> If the control with the items is `wxListCtrl` (and not `wxDataViewCtrl`), then this looks very similar to #24443 and if FileZilla uses `wxEVT_IDLE` for refreshing, this must definitely be the same thing (and it still might be even if it doesn't).
> 
> Unfortunately I have no idea about the artifacts at the bottom.

For your information, I downloaded the Filezilla source code and made a search into the project. There are several `wxListCtrl`, whereas `wxDataViewCtrl` is not preset and I can't find `wxEVT_IDLE` either.

While searching `wxListCtrl`, I found out some things that might be interesting.

In `src/interface/filelistctrl.h` there is:

```c++
#ifndef __WXMSW__
	// Generic wxListCtrl does not support wxLIST_STATE_DROPHILITED, emulate it
	wxListItemAttr m_dropHighlightAttribute;
#endif
```

In `src/interface/listctrlex.cpp` there's a comment at the bottom of a function saying `generic wxListCtrl needs manual refresh`:

```c++
void wxListCtrlEx::ShowColumnEditor()
{
	CColumnEditDialog dlg;

	if (!dlg.Create(this)) {
		return;
	}
	
	dlg.order_.resize(m_columnInfo.size());
	for (size_t j = 0; j < m_columnInfo.size(); ++j) {
		for (size_t i = 0; i < m_columnInfo.size(); ++i) {
			if (m_columnInfo[i].order != j) {
				continue;
			}
			dlg.order_[j] = i;
			dlg.list_box_->Append(m_columnInfo[i].name);
			if (m_columnInfo[i].shown) {
				dlg.list_box_->Check(j);
			}
		}
	}
	wxASSERT(dlg.list_box_->GetCount() == m_columnInfo.size());

	dlg.GetSizer()->Fit(&dlg);

	if (dlg.ShowModal() != wxID_OK) {
		return;
	}

	for (size_t i = 0; i < m_columnInfo.size(); ++i) {
		int col = dlg.order_[i];
		bool isChecked = dlg.list_box_->IsChecked(i);
		if (!isChecked && !col) {
			isChecked = true;
			wxMessageBoxEx(_("The filename column cannot be hidden."));
		}
		MoveColumn(col, i);
		if (m_columnInfo[col].shown != isChecked) {
			ShowColumn(col, isChecked);
		}
	}

	// Generic wxListCtrl needs manual refresh
	Refresh();
}
```

**Comment by Digitalone1** [comment]
For what it's worth, I tried to compile the framework removing the check reported [here](https://github.com/wxWidgets/wxWidgets/issues/24443#issuecomment-2567205646), but I still get the same issue in Filezilla.

I also tried to use amule. It does not present the same issue as Filezilla since the lists are "stable" (whereas Filezilla prints some messages on them that gets stuck/not refreshed). But they are still broken because if you scroll them, sometimes they are not refreshed.

It's unfortunate that those widgets are barely usable under Wayland. Most known desktop environments on Linux use it as the default session and having to go back to X11 for wxWidgets apps seems a very bad sign.

**Comment by vadz** [comment]
I already wrote that I agree that it's a bad bug, the problem is that

1. I still don't have any simple way to reproduce it — patching listctrl sample to allow doing this using the code from FileZilla or aMule would be most welcome.
2. Even in my own program where I can reproduce it, I can only find a workaround (which doesn't even work for you) that I don't really understand.

So yes, more work is needed here and any help is welcome.

**Comment by paulcor** [comment]
I can reproduce with Filezilla. I think it's caused by `Update()`.

**Comment by paulcor** [comment]
The problem is caused by `wxWindow::Update()`, but I don't know exactly why. AFAICT, here is what happens:
1. a `wxListCtrl` is modified, and `wxStatusBar::SetStatusText()` is called
2. `wxStatusBar` calls `Update()`, forcing a paint event for the list control (along with many others)
3. the list control is "dirty", so it doesn't paint, but arranges for a `Refresh()` call later from an idle handler (this is a bit of a WTF, IMO)
4. the `Update()` painting finishes, seems to (mostly?) work, sometime later an idle occurs, `Refresh()` is called on the list control
5. the list control gets a regular paint event, the proper drawing seems to happen, the update region looks fine, but it all just goes to bit heaven; nothing makes it to the screen

I have not been able to make a simple test case. The Filezilla reproducer is complicated, with a tree control, a list control and a second status bar all buried in 5 levels of splitter window. The tree control also has drawing problems, which I think are related.

Frankly I'm surprised `Update()` seems to work at all under Wayland, and without knowing exactly why it causes a problem, the only fix appears to be removing the `Update()` from `wxStatusBar` and/or disabling `Update()` entirely for Wayland.

**Comment by vadz** [comment]
Thanks for debugging this! In my app I have a status bar too, but I had never realized it could play a role. I do remember that during my debugging I saw paint events which never resulted in any output and I couldn't find any reason for it neither.

I'll try to check if removing `Update()` solves the problem in my case too. I wonder if we could somehow pause repainting of all the other windows than the one on which it is called to still allow to show some message in the status bar immediately without affecting anything else?

I can also look at removing "dirty" flag from `wxListCtrl`...

**Comment by vadz** [comment]
*Tl;DR*: @Digitalone1 Could you please try #25204 and let me know if this fixes the issue for you too?

I can confirm that changing `Update()` to do nothing also fixes the problem in my application, so I think we should make it do nothing when using Wayland — this is obviously not ideal, but arguably better than what we have now.

Besides, I don't think it works under Wayland anyhow currently, at least my test application:

```cpp
#include <wx/app.h>
#include <wx/button.h>
#include <wx/datetime.h>
#include <wx/frame.h>
#include <wx/log.h>

class TestApp : public wxApp {
public:
    bool OnInit() override {
        auto* frame = new wxFrame(nullptr, wxID_ANY, "Hello, World!");
        frame->CreateStatusBar();

        auto* b = new wxButton(frame, wxID_ANY, "Click me!");
        b->Bind(wxEVT_BUTTON, [frame](wxCommandEvent&) {
            frame->SetStatusText(wxString::Format("Button clicked at %s", wxDateTime::UNow().FormatTime()));
            wxSleep(2);
            frame->SetStatusText(wxString::Format("Woke up at %s", wxDateTime::UNow().FormatTime()));
        });

        frame->Show(true);

        return true;
    }
};

wxIMPLEMENT_APP(TestApp);
```

doesn't show "Button clicked" in the status bar when using Wayland (but does when using X).

**Comment by Digitalone1** [comment]
@vadz  Sorry, I'm not using Filezilla anymore, so I can't test it. Thanks for your work anyway.

**Comment by vadz** [comment]
I probably shouldn't be spending so much time on this as I don't think I can really fix `Update()` with Wayland, but I couldn't help at least trying to understand what would it take to do it because it seems like it should really be possible: after all, with Wayland, nothing prevents the application from committing the surface at any time, it doesn't _need_ to wait for `wl_callback_listener::done` event, even though this is usually done.

But by reading GTK code I see that, other than at window creation, `wl_surface_commit()` is only called from the window handler of `after-paint` signal. And this one, in turn, is generated only by `gdk_frame_clock_paint_idle()` which is only called when the timeout added by the internal `maybe_start_idle ()` expires. So the only way this could work from `Update()`, AFAICS, is to somehow force Glib to call timeout handlers immediately, but I don't think this is possible.

So unless @paulcor has some magic solution for this, I can only suggest merging #25204 to which I will add a documentation change saying that `Update()` does nothing when using Wayland.
